```
experimental.sol:1:1: Warning: Experimental features are turned on. Do not use experimental features on live deployments.
pragma experimental __test;
^-------------------------^
experimental.sol:2:1: Error: Duplicate experimental feature name.
pragma experimental __test;
^-------------------------^
```

None of the tests helpers `CHECK_ERROR`, `CHECK_ERROR_ALLOW_MULTI`, `CHECK_WARNING_ALLOW_MULTI` can properly catch the "Duplicate" message. Any ideas? I guess the reason is the warning is emitted first.

```
/Users/alex/Projects/solidity/test/libsolidity/SolidityNameAndTypeResolution.cpp:6595: error: in "SolidityNameAndTypeResolution/experimental_pragma": check err.type() == (Error::Type::SyntaxError) has failed
Expected message "Duplicate experimental feature name." but found "Experimental features are turned on. Do not use experimental features on live deployments.".
/Users/alex/Projects/solidity/test/libsolidity/SolidityNameAndTypeResolution.cpp:6595: error: in "SolidityNameAndTypeResolution/experimental_pragma": check searchErrorMessage(err, ("Duplicate experimental feature name.")) has failed
```

Alternatively we could allow duplicates.

--- a/libsolidity/ast/ExperimentalFeatures.h
+++ b/libsolidity/ast/ExperimentalFeatures.h
@@ -27,12 +27,19 @@ namespace dev
 namespace solidity
 {
 
-enum class ExperimentalFeature {
-	Test
+enum class ExperimentalFeature
+{
+	Test,
+	TestOnlyAnalysis
+};
+
+static const std::map<ExperimentalFeature, bool> ExperimentalFeatureOnlyAnalysis = {
+	{ ExperimentalFeature::TestOnlyAnalysis, true },
 };
 
 static const std::map<std::string, ExperimentalFeature> ExperimentalFeatureNames = {
 	{ "__test", ExperimentalFeature::Test },
+	{ "__testOnlyAnalysis", ExperimentalFeature::TestOnlyAnalysis },
 };
 
 }

--- a/libsolidity/ast/ExperimentalFeatures.h
+++ b/libsolidity/ast/ExperimentalFeatures.h
@@ -27,9 +27,13 @@ namespace dev
 namespace solidity
 {
 
-enum class ExperimentalFeature {};
+enum class ExperimentalFeature {
+	Test
+};
 
-static const std::map<std::string, ExperimentalFeature> ExperimentalFeatureNames = {};
+static const std::map<std::string, ExperimentalFeature> ExperimentalFeatureNames = {
+	{ "__test", ExperimentalFeature::Test },
+};
 
 }
 }

--- a/libsolidity/interface/CompilerStack.cpp
+++ b/libsolidity/interface/CompilerStack.cpp
@@ -632,6 +632,17 @@ string CompilerStack::absolutePath(string const& _path, string const& _reference
 	return result.generic_string();
 }
 
+namespace
+{
+bool onlySafeExperimentalFeaturesActivated(set<ExperimentalFeature> const& features)
+{
+	for (auto const feature: features)
+		if (!ExperimentalFeatureOnlyAnalysis.count(feature))
+			return false;
+	return true;
+}
+}
+
 void CompilerStack::compileContract(
 	ContractDefinition const& _contract,
 	map<ContractDefinition const*, eth::Assembly const*>& _compiledContracts
@@ -649,16 +660,23 @@ void CompilerStack::compileContract(
 	shared_ptr<Compiler> compiler = make_shared<Compiler>(m_optimize, m_optimizeRuns);
 	Contract& compiledContract = m_contracts.at(_contract.fullyQualifiedName());
 	string metadata = createMetadata(compiledContract);
-	bytes cborEncodedMetadata =
-		// CBOR-encoding of {"bzzr0": dev::swarmHash(metadata)}
-		bytes{0xa1, 0x65, 'b', 'z', 'z', 'r', '0', 0x58, 0x20} +
-		dev::swarmHash(metadata).asBytes();
-	if (!_contract.sourceUnit().annotation().experimentalFeatures.empty())
+	bytes cborEncodedHash =
+		// CBOR-encoding of the key "bzzr0"
+		bytes{0x65, 'b', 'z', 'z', 'r', '0'}+
+		// CBOR-encoding of the hash
+		bytes{0x58, 0x20} + dev::swarmHash(metadata).asBytes();
+	bytes cborEncodedMetadata;
+	if (onlySafeExperimentalFeaturesActivated(_contract.sourceUnit().annotation().experimentalFeatures))
 		cborEncodedMetadata =
-		// CBOR-encoding of {"bzzr0": dev::swarmHash(metadata), "experimental": true}
-		bytes{0xa2, 0x65, 'b', 'z', 'z', 'r', '0', 0x58, 0x20} +
-		dev::swarmHash(metadata).asBytes() +
-		bytes{0x6c, 'e', 'x', 'p', 'e', 'r', 'i', 'm', 'e', 'n', 't', 'a', 'l', 0xf5};
+			// CBOR-encoding of {"bzzr0": dev::swarmHash(metadata)}
+			bytes{0xa1} +
+			cborEncodedHash;
+	else
+		cborEncodedMetadata =
+			// CBOR-encoding of {"bzzr0": dev::swarmHash(metadata), "experimental": true}
+			bytes{0xa2} +
+			cborEncodedHash +
+			bytes{0x6c, 'e', 'x', 'p', 'e', 'r', 'i', 'm', 'e', 'n', 't', 'a', 'l', 0xf5};
 	solAssert(cborEncodedMetadata.size() <= 0xffff, "Metadata too large");
 	// 16-bit big endian length
 	cborEncodedMetadata += toCompactBigEndian(cborEncodedMetadata.size(), 2);

--- a/test/libsolidity/Metadata.cpp
+++ b/test/libsolidity/Metadata.cpp
@@ -38,6 +38,7 @@ BOOST_AUTO_TEST_CASE(metadata_stamp)
 	// Check that the metadata stamp is at the end of the runtime bytecode.
 	char const* sourceCode = R"(
 		pragma solidity >=0.0;
+		pragma experimental __testOnlyAnalysis;
 		contract test {
 			function g(function(uint) external returns (uint) x) {}
 		}

--- a/test/libsolidity/SolidityNameAndTypeResolution.cpp
+++ b/test/libsolidity/SolidityNameAndTypeResolution.cpp
@@ -6584,15 +6584,15 @@ BOOST_AUTO_TEST_CASE(experimental_pragma)
 		pragma experimental unsupportedName unsupportedName;
 	)";
 	CHECK_ERROR(text, SyntaxError, "Stray arguments.");
+	text = R"(
+		pragma experimental __test;
+	)";
+	CHECK_WARNING(text, "Experimental features are turned on. Do not use experimental features on live deployments.");
 //	text = R"(
-//		pragma experimental supportedName;
+//		pragma experimental __test;
+//		pragma experimental __test;
 //	)";
-//	CHECK_WARNING(text, "Experimental features are turned on. Do not use experimental features on live deployments.");
-//	text = R"(
-//		pragma experimental supportedName;
-//		pragma experimental supportedName;
-//	)";
-//	CHECK_ERROR(text, SyntaxError, "Duplicate experimental feature name.");
+//	CHECK_ERROR_ALLOW_MULTI(text, SyntaxError, "Duplicate experimental feature name.");
 }
 
 BOOST_AUTO_TEST_SUITE_END()

